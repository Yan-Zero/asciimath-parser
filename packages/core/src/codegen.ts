import type { ChildNode, MatrixNode, RootNode } from './parser'
import { AlignDirection, NodeTypes } from './parser'

function getArrayBoundary(node: MatrixNode) {
  const div = node.dividerIndices
  let beginArray = '\\begin{array}'
  let ch = 'c'
  if (node.alignment === AlignDirection.Left)
    ch = 'l'

  if (div.length) {
    const divMax = div[div.length - 1]
    for (let i = div.length - 1; i >= 1; i--)
      div[i] -= div[i - 1]

    beginArray += '{'
    for (let i = 0; i < div.length; i++)
      beginArray += `${''.padEnd(div[i], ch)}|`

    // MathJax would complain if the array env arg
    // is not consistent with the elements of the matrix.
    const maxCol = Math.max(...node.params.map(i => i.length))
    beginArray += `${''.padEnd(maxCol - divMax, ch)}}`
  }
  else {
    const maxCol = Math.max(...node.params.map(c => c.length))
    beginArray += '{' + ''.padEnd(maxCol, ch) + '}'
  }
  return [beginArray, '\\end{array}']
}

function codegen(node: ChildNode | RootNode): string {
  switch (node.type) {
    case NodeTypes.Const: {
      return node.tex
    }
    case NodeTypes.Root: {
      let res = node.body.map(codegen).join(' ')
      if (node.body.find(n => n.type === NodeTypes.Const && n.value === '&'))
        res = `\\begin{aligned}${res}\\end{aligned}`
      return res
    }
    case NodeTypes.Flat: {
      return node.body.map(codegen).join(' ')
    }
    case NodeTypes.Matrix: {
      // const [beginMatrix, endMatrix] = getMatrixBoundary(node)
      const [arrayBegin, arrayEnd] = getArrayBoundary(node)
      return [
        node.lparen,
        // beginMatrix,
        arrayBegin,
        node.params.map(i => i.map(codegen).join('&')).join('\\\\'),
        '\\\\',
        arrayEnd,
        // endMatrix,
        node.rparen,
      ].join(' ')
    }
    case NodeTypes.ParamOne: {
      return node.tex.replace('$1', codegen(node.params))
    }
    case NodeTypes.ParamTwo: {
      return node.tex.replace('$1', codegen(node.params[0])).replace('$2', codegen(node.params[1]))
    }
  }
}

export {
  codegen,
}
