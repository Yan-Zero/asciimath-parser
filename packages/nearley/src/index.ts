import Nearley from 'nearley'
import initGrammar from './grammar.js'
import initLexer from './lexer.js'
import type { Symbols } from './symbols.js'
import initSymbols from './symbols.js'
import initGenerator from './to-tex.js'

export type Ast = any
type ReplaceLaw = [RegExp | string, string | ((substring: string, ...args: any[]) => string)]

interface AsciiMathConfig {
  /**
   * @default true
   * enable displayMode in KaTeX
   */
  display?: boolean
  /**
   * @default false
   * throws error on parse
   */
  throws?: boolean
  /**
   * Add custom symbols.
   *
   * For example:
   * {
   *   keyword: {
   *     dx: { tex: '{\\text{d}x}' },
   *     dy: { tex: '{\\text{d}y}' },
   *     dz: { tex: '{\\text{d}z}' },
   *     dt: { tex: '{\\text{d}t}' },
   *     ee: { tex: '\\text{e}' },
   *     ii: { tex: '\\text{i}' },
   *   }
   * }
   */
  symbols?: Symbols
  /**
   * Replace target expressions before tokenizing
   *
   * For example:
   * [
   *   [/&#(x?[0-9a-fA-F]+);/g, (match, $1) =>
   *     String.fromCodePoint($1[0] === 'x' ? '0' + $1 : $1)
   *   ],
   *   ...
   * ]
   */
  replaceBeforeTokenizing?: ReplaceLaw[]
}

type RestrictedAmConfig = Required<AsciiMathConfig>

function resolveConfig(config?: AsciiMathConfig): RestrictedAmConfig {
  return {
    display: config?.display ?? true,
    throws: config?.throws ?? false,
    symbols: {
      keyword: {
        dx: { tex: '{\\text{d}x}' },
        dy: { tex: '{\\text{d}y}' },
        dz: { tex: '{\\text{d}z}' },
        dt: { tex: '{\\text{d}t}' },
        ee: { tex: '\\text{e}' },
        ii: { tex: '\\text{i}' },
      },
      ...config?.symbols,
    },
    replaceBeforeTokenizing: [
      // html entity
      [/&#(x?[0-9a-fA-F]+);/g, (_match, $1) =>
        String.fromCodePoint($1[0] === 'x' ? `0${$1}` : $1),
      ],
      ...(config?.replaceBeforeTokenizing || []),
    ],
  }
}

class AsciiMath {
  public display: boolean
  public throws: boolean
  public replaceBeforeTokenizing: ReplaceLaw[]
  public lexer: Nearley.Lexer
  public parser: Nearley.Parser
  private genTex: ((ast: Ast) => string)
  private initState: { [key: string]: any; lexerState: Nearley.LexerState }
  constructor(config?: AsciiMathConfig) {
    const { display, throws, symbols, replaceBeforeTokenizing } = resolveConfig(config)
    this.display = display
    this.throws = throws
    this.replaceBeforeTokenizing = replaceBeforeTokenizing
    const allSymbols = initSymbols(symbols)
    const lexer = this.lexer = initLexer(allSymbols)
    const grammar = initGrammar(lexer)
    const compiledGrammar = Nearley.Grammar.fromCompiled(grammar)
    this.parser = new Nearley.Parser(compiledGrammar)
    this.initState = this.parser.save()
    this.genTex = initGenerator(allSymbols)
  }

  toTex(code: string): string {
    this.parser.restore(this.initState)
    try {
      code = this.replaceBeforeTokenizing.reduce((prev, curLaw) => {
        // nodejs 14 没有 replaceAll 方法, 不要用 replaceAll
        return prev.replace(new RegExp(curLaw[0], 'g'), curLaw[1] as any)
      }, code)
      this.parser.feed(code)
      if (this.parser.results.length === 0) {
        throw new Error('unexpected end of input')
      }
      else if (this.parser.results.length > 1) {
        console.error(this.parser.results)
        throw new Error('ambiguous parse')
      }
      let res = this.genTex(this.parser.results)
      if (this.display)
        res = `\\displaystyle{ ${res} }`
      return res
    }
    catch (e) {
      if (this.throws)
        throw e
      console.error(e)
      const message = String(e)
      let index = message.indexOf(' Instead, I was expecting to see one of the following:')
      if (index > -1) {
        return this.genTex({
          type: 'opOA',
          value: 'verb',
          $1: { value: message.slice(0, index) },
        })
      }

      index = message.indexOf(':\n')
      const end = index > -1 ? index : undefined
      return `\\text{${message.slice(0, end)}}`
    }
  }
}

export {
  AsciiMath,
  AsciiMathConfig,
}
