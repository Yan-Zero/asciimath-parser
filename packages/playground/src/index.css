@import url('https://fonts.googleapis.com/css2?family=JetBrains+Mono&display=swap');

:root {
  font-family: Inter, Avenir, Helvetica, Arial, sans-serif;
  font-size: 16px;
  line-height: 24px;
  font-weight: 400;

  color-scheme: light dark;
  color: rgba(255, 255, 255, 0.87);
  background-color: #242424;

  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: 100%;
}

a {
  font-weight: 500;
  color: #646cff;
  text-decoration: inherit;
}
a:hover {
  color: #535bf2;
}

body {
  margin: 0;
  min-width: 320px;
  min-height: 100vh;
}

h1 {
  font-size: 3.2em;
  line-height: 1.1;
  margin-block-end: 0.25rem;
}

h2 {
  margin-top: 3rem;
}

h3 {
  margin-top: 2rem;
}

button {
  border-radius: 8px;
  border: 1px solid transparent;
  padding: 0.6em 1.2em;
  font-size: 1em;
  font-weight: 500;
  font-family: inherit;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.25s;
  display: flex;
  align-items: center;
  justify-content: center;
}
button:hover {
  border-color: #646cff;
}
button:focus,
button:focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

@media (prefers-color-scheme: light) {
  :root {
    color: #213547;
    background-color: #ffffff;
  }
  a:hover {
    color: #747bff;
  }
  button {
    background-color: #f9f9f9;
  }
}

td {
  padding: 0.5rem;
  white-space: nowrap;
}

table {
  width: 100%;
  border-top: 2px solid;
  border-bottom: 2px solid;
}

.tc {
  text-align: center;
}

.tex {
  background-color: #7f7f7f33;
}

main {
  max-width: 800px;
  margin: 0 auto;
  padding: 1rem;
}

.space-x-1 > :not([hidden]) ~ :not([hidden]) {
  --tw-space-x-reverse: 0;
  margin-right: calc(0.25rem * var(--tw-space-x-reverse));
  margin-left: calc(0.25rem * calc(1 - var(--tw-space-x-reverse)));
}

.flex {
  display: flex;
}
.w-full {
  width: 100%;
}
.mb-4 {
  margin-bottom: 1rem;
}

code, pre {
  font-family: 'JetBrains Mono', 'Courier New', Courier, monospace;
  font-size: 0.8rem;
}

code {
  background-color: #7f7f7f11;
  padding: 0.2rem;
  border-radius: 0.2rem;
}

pre {
  background-color: #7f7f7f11;
  padding: 0.75em;
  border-radius: 0.2rem;
  overflow: auto;
}