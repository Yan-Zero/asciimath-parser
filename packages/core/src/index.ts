import { codegen } from './codegen'
import { parser } from './parser'
import type { Trie } from './trie'
import { createTrie } from './trie'

type ReplaceLaw = [RegExp | string, string | ((substring: string, ...args: any[]) => string)]

interface AsciiMathConfig {
  /**
   * @default true
   * enable displayMode in KaTeX
   */
  display?: boolean
  /**
   * Translate custom keywords into LaTeX expressions
   *
   * For example:
   * [
   *   ['dx', '\text{d}x'],
   *   ['dy', '\text{d}y']
   * ]
   */
  extConst?: Array<[string, string]>
  /**
   * Replace target expressions before tokenizing
   *
   * For example:
   * [
   *   [/&#(x?[0-9a-fA-F]+);/g, (match, $1) =>
   *     String.fromCodePoint($1[0] === 'x' ? '0' + $1 : $1)
   *   ],
   *   ...
   * ]
   */
  replaceBeforeTokenizing?: ReplaceLaw[]
}

type RestrictedAmConfig = Required<AsciiMathConfig>

function resolveConfig(config?: AsciiMathConfig): RestrictedAmConfig {
  const defaultConfig: RestrictedAmConfig = {
    display: true,
    extConst: [
      ['dx', '{\\text{d}x}'],
      ['dy', '{\\text{d}y}'],
      ['dz', '{\\text{d}z}'],
      ['dt', '{\\text{d}t}'],
      ['#', '\\displaystyle'],
    ],
    replaceBeforeTokenizing: [
      [/&#(x?[0-9a-fA-F]+);/g, (_match, $1) =>
        String.fromCodePoint($1[0] === 'x' ? `0${$1}` : $1),
      ],
    ],
  }
  if (typeof config?.display !== 'undefined')
    defaultConfig.display = config?.display
  if (config?.extConst?.length)
    defaultConfig.extConst.push(...config.extConst)
  if (config?.replaceBeforeTokenizing?.length)
    defaultConfig.replaceBeforeTokenizing.push(...config.replaceBeforeTokenizing)

  return defaultConfig
}

class AsciiMath {
  private trie: Trie
  private display: boolean
  private replaceLaws: ReplaceLaw[]
  constructor(config?: AsciiMathConfig) {
    const { display, extConst, replaceBeforeTokenizing: replaceBeforeParsing } = resolveConfig(config)
    this.trie = createTrie({ extConst })
    this.display = display
    this.replaceLaws = replaceBeforeParsing
  }

  toTex(code: string): string {
    try {
      code = this.replaceLaws.reduce((prev, curLaw) => {
      // in fact the judgement is no use...
        if (typeof curLaw[1] === 'function')
          return prev.replace(curLaw[0], curLaw[1])
        else
          return prev.replace(curLaw[0], curLaw[1])
      }, code)
      let res = codegen(parser(this.trie.tryParsingAll(code)))
      if (this.display)
        res = `\\displaystyle{ ${res} }`
      return res
    }
    catch (e) {
      return `\\text{${String(e)}}`
    }
  }
}

export {
  AsciiMath,
  AsciiMathConfig,
}
