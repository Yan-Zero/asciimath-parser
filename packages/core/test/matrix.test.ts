import { describe, expect, it } from 'vitest'
import { AsciiMath } from '../src'
import { codegen } from '../src/codegen'
import { parser } from '../src/parser'
import { createTrie } from '../src/trie'

const code = `
[1, 0, cdots, 0;
s_1, b_11, cdots, b_(1n);
vdots, vdots, , vdots;
s_n, b_(n1), cdots, b_(n n)]`

describe('matrix with empty el', () => {
//   const am = new AsciiMath({ display: false })
  it('should generate ast', () => {
    const trie = createTrie()
    const tokens = trie.tryParsingAll(code)
    const ast = parser(tokens)
    expect(tokens).toMatchSnapshot()
    expect(ast).toMatchSnapshot()
    expect(codegen(ast)).toMatchInlineSnapshot('"\\\\left[ \\\\begin{array}{cccc} 1 & 0 & \\\\cdots & 0\\\\\\\\s _{ 1 } & b _{ 11 } & \\\\cdots & b _{ 1 n }\\\\\\\\\\\\vdots & \\\\vdots &  & \\\\vdots\\\\\\\\s _{ n } & b _{ n 1 } & \\\\cdots & b _{ n n } \\\\end{array} \\\\right]"')
  })
})

describe('centered matrix with no braces', () => {
  it('should generate centered matrix', () => {
    const trie = createTrie()
    const tokens = trie.tryParsingAll('{: a, b; c, d :}')
    const ast = parser(tokens)
    expect(tokens).toMatchSnapshot()
    expect(ast).toMatchSnapshot()
    expect(codegen(ast)).toMatchInlineSnapshot('"\\\\left. \\\\begin{array}{cc} a & b\\\\\\\\c & d \\\\end{array} \\\\right."')
  })

  it('should remove () of matrix in superscript', () => {
    const am = new AsciiMath({ display: false })
    expect(am.toTex('sum_(a;b)')).toMatchInlineSnapshot('"\\\\sum _{  \\\\begin{array}{c} a\\\\\\\\b \\\\end{array}  }"')
  })

  it('should generate left aligned matrix', () => {
    const am = new AsciiMath({ display: false })
    expect(am.toTex('|x|={x, if x > 0; -x, otherwise :}')).toMatchInlineSnapshot('"\\\\mid x |= \\\\left\\\\lbrace \\\\begin{array}{ll} x & \\\\text{if}\\\\quad x > 0\\\\\\\\- x & \\\\text{otherwise}\\\\quad \\\\end{array} \\\\right."')
  })
})

describe('grid like matrix', () => {
  it('should generate grid like matrix', () => {
    const trie = createTrie()
    const tokens = trie.tryParsingAll('{: | hline a, b;hline c|d|; hline :}')
    const ast = parser(tokens)
    expect(tokens).toMatchSnapshot()
    expect(ast).toMatchSnapshot()
    expect(codegen(ast)).toMatchSnapshot()
  })
})
