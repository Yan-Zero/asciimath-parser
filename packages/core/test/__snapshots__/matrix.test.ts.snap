// Vitest Snapshot v1

exports[`centered matrix with no braces > should generate centered matrix 1`] = `
[
  {
    "current": 2,
    "isKeyWord": true,
    "tex": ".",
    "type": "LParen",
    "value": "{:",
  },
  {
    "current": 4,
    "isKeyWord": false,
    "tex": "a",
    "type": "StringLiteral",
    "value": "a",
  },
  {
    "current": 5,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 7,
    "isKeyWord": false,
    "tex": "b",
    "type": "StringLiteral",
    "value": "b",
  },
  {
    "current": 8,
    "isKeyWord": true,
    "tex": ";",
    "type": "Split",
    "value": ";",
  },
  {
    "current": 10,
    "isKeyWord": false,
    "tex": "c",
    "type": "StringLiteral",
    "value": "c",
  },
  {
    "current": 11,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 13,
    "isKeyWord": false,
    "tex": "d",
    "type": "StringLiteral",
    "value": "d",
  },
  {
    "current": 16,
    "isKeyWord": true,
    "tex": ".",
    "type": "RParen",
    "value": ":}",
  },
]
`;

exports[`centered matrix with no braces > should generate centered matrix 2`] = `
{
  "body": [
    {
      "alignment": "c",
      "dividerIndices": [],
      "lparen": "\\\\left.",
      "params": [
        [
          {
            "body": [
              {
                "tex": "a",
                "type": "Const",
                "value": "a",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "tex": "b",
                "type": "Const",
                "value": "b",
              },
            ],
            "type": "Flat",
          },
        ],
        [
          {
            "body": [
              {
                "tex": "c",
                "type": "Const",
                "value": "c",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "tex": "d",
                "type": "Const",
                "value": "d",
              },
            ],
            "type": "Flat",
          },
        ],
      ],
      "rparen": "\\\\right.",
      "type": "Matrix",
    },
  ],
  "type": "Root",
}
`;

exports[`grid like matrix > should generate grid like matrix 1`] = `
[
  {
    "current": 2,
    "isKeyWord": true,
    "tex": ".",
    "type": "LParen",
    "value": "{:",
  },
  {
    "current": 4,
    "isKeyWord": true,
    "tex": "|",
    "type": "Paren",
    "value": "|",
  },
  {
    "current": 10,
    "isKeyWord": true,
    "tex": "\\\\hline",
    "type": "Const",
    "value": "hline",
  },
  {
    "current": 12,
    "isKeyWord": false,
    "tex": "a",
    "type": "StringLiteral",
    "value": "a",
  },
  {
    "current": 13,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 15,
    "isKeyWord": false,
    "tex": "b",
    "type": "StringLiteral",
    "value": "b",
  },
  {
    "current": 16,
    "isKeyWord": true,
    "tex": ";",
    "type": "Split",
    "value": ";",
  },
  {
    "current": 21,
    "isKeyWord": true,
    "tex": "\\\\hline",
    "type": "Const",
    "value": "hline",
  },
  {
    "current": 23,
    "isKeyWord": false,
    "tex": "c",
    "type": "StringLiteral",
    "value": "c",
  },
  {
    "current": 24,
    "isKeyWord": true,
    "tex": "|",
    "type": "Paren",
    "value": "|",
  },
  {
    "current": 25,
    "isKeyWord": false,
    "tex": "d",
    "type": "StringLiteral",
    "value": "d",
  },
  {
    "current": 26,
    "isKeyWord": true,
    "tex": "|",
    "type": "Paren",
    "value": "|",
  },
  {
    "current": 27,
    "isKeyWord": true,
    "tex": ";",
    "type": "Split",
    "value": ";",
  },
  {
    "current": 33,
    "isKeyWord": true,
    "tex": "\\\\hline",
    "type": "Const",
    "value": "hline",
  },
  {
    "current": 36,
    "isKeyWord": true,
    "tex": ".",
    "type": "RParen",
    "value": ":}",
  },
]
`;

exports[`grid like matrix > should generate grid like matrix 2`] = `
{
  "body": [
    {
      "alignment": "c",
      "dividerIndices": [
        0,
        1,
        2,
      ],
      "lparen": "\\\\left.",
      "params": [
        [
          {
            "body": [
              {
                "tex": "\\\\hline",
                "type": "Const",
                "value": "hline",
              },
              {
                "tex": "a",
                "type": "Const",
                "value": "a",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "tex": "b",
                "type": "Const",
                "value": "b",
              },
            ],
            "type": "Flat",
          },
        ],
        [
          {
            "body": [
              {
                "tex": "\\\\hline",
                "type": "Const",
                "value": "hline",
              },
              {
                "tex": "c",
                "type": "Const",
                "value": "c",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "tex": "d",
                "type": "Const",
                "value": "d",
              },
            ],
            "type": "Flat",
          },
        ],
        [
          {
            "body": [
              {
                "tex": "\\\\hline",
                "type": "Const",
                "value": "hline",
              },
            ],
            "type": "Flat",
          },
        ],
      ],
      "rparen": "\\\\right.",
      "type": "Matrix",
    },
  ],
  "type": "Root",
}
`;

exports[`grid like matrix > should generate grid like matrix 3`] = `"\\\\left. \\\\begin{array}{|c|c|} \\\\hline a & b\\\\\\\\\\\\hline c & d\\\\\\\\\\\\hline \\\\end{array} \\\\right."`;

exports[`matrix with empty el > should generate ast 1`] = `
[
  {
    "current": 2,
    "isKeyWord": true,
    "tex": "[",
    "type": "LParen",
    "value": "[",
  },
  {
    "current": 3,
    "isKeyWord": false,
    "tex": "1",
    "type": "NumberLiteral",
    "value": "1",
  },
  {
    "current": 4,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 6,
    "isKeyWord": false,
    "tex": "0",
    "type": "NumberLiteral",
    "value": "0",
  },
  {
    "current": 7,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 13,
    "isKeyWord": true,
    "tex": "\\\\cdots",
    "type": "Const",
    "value": "cdots",
  },
  {
    "current": 14,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 16,
    "isKeyWord": false,
    "tex": "0",
    "type": "NumberLiteral",
    "value": "0",
  },
  {
    "current": 17,
    "isKeyWord": true,
    "tex": ";",
    "type": "Split",
    "value": ";",
  },
  {
    "current": 19,
    "isKeyWord": false,
    "tex": "s",
    "type": "StringLiteral",
    "value": "s",
  },
  {
    "current": 20,
    "isKeyWord": true,
    "tex": "_{ $1 }",
    "type": "OperatorSup",
    "value": "_",
  },
  {
    "current": 21,
    "isKeyWord": false,
    "tex": "1",
    "type": "NumberLiteral",
    "value": "1",
  },
  {
    "current": 22,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 24,
    "isKeyWord": false,
    "tex": "b",
    "type": "StringLiteral",
    "value": "b",
  },
  {
    "current": 25,
    "isKeyWord": true,
    "tex": "_{ $1 }",
    "type": "OperatorSup",
    "value": "_",
  },
  {
    "current": 27,
    "isKeyWord": false,
    "tex": "11",
    "type": "NumberLiteral",
    "value": "11",
  },
  {
    "current": 28,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 34,
    "isKeyWord": true,
    "tex": "\\\\cdots",
    "type": "Const",
    "value": "cdots",
  },
  {
    "current": 35,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 37,
    "isKeyWord": false,
    "tex": "b",
    "type": "StringLiteral",
    "value": "b",
  },
  {
    "current": 38,
    "isKeyWord": true,
    "tex": "_{ $1 }",
    "type": "OperatorSup",
    "value": "_",
  },
  {
    "current": 39,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 40,
    "isKeyWord": false,
    "tex": "1",
    "type": "NumberLiteral",
    "value": "1",
  },
  {
    "current": 41,
    "isKeyWord": false,
    "tex": "n",
    "type": "StringLiteral",
    "value": "n",
  },
  {
    "current": 42,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
  {
    "current": 43,
    "isKeyWord": true,
    "tex": ";",
    "type": "Split",
    "value": ";",
  },
  {
    "current": 49,
    "isKeyWord": true,
    "tex": "\\\\vdots",
    "type": "Const",
    "value": "vdots",
  },
  {
    "current": 50,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 56,
    "isKeyWord": true,
    "tex": "\\\\vdots",
    "type": "Const",
    "value": "vdots",
  },
  {
    "current": 57,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 59,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 65,
    "isKeyWord": true,
    "tex": "\\\\vdots",
    "type": "Const",
    "value": "vdots",
  },
  {
    "current": 66,
    "isKeyWord": true,
    "tex": ";",
    "type": "Split",
    "value": ";",
  },
  {
    "current": 68,
    "isKeyWord": false,
    "tex": "s",
    "type": "StringLiteral",
    "value": "s",
  },
  {
    "current": 69,
    "isKeyWord": true,
    "tex": "_{ $1 }",
    "type": "OperatorSup",
    "value": "_",
  },
  {
    "current": 70,
    "isKeyWord": false,
    "tex": "n",
    "type": "StringLiteral",
    "value": "n",
  },
  {
    "current": 71,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 73,
    "isKeyWord": false,
    "tex": "b",
    "type": "StringLiteral",
    "value": "b",
  },
  {
    "current": 74,
    "isKeyWord": true,
    "tex": "_{ $1 }",
    "type": "OperatorSup",
    "value": "_",
  },
  {
    "current": 75,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 76,
    "isKeyWord": false,
    "tex": "n",
    "type": "StringLiteral",
    "value": "n",
  },
  {
    "current": 77,
    "isKeyWord": false,
    "tex": "1",
    "type": "NumberLiteral",
    "value": "1",
  },
  {
    "current": 78,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
  {
    "current": 79,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 85,
    "isKeyWord": true,
    "tex": "\\\\cdots",
    "type": "Const",
    "value": "cdots",
  },
  {
    "current": 86,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 88,
    "isKeyWord": false,
    "tex": "b",
    "type": "StringLiteral",
    "value": "b",
  },
  {
    "current": 89,
    "isKeyWord": true,
    "tex": "_{ $1 }",
    "type": "OperatorSup",
    "value": "_",
  },
  {
    "current": 90,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 91,
    "isKeyWord": false,
    "tex": "n",
    "type": "StringLiteral",
    "value": "n",
  },
  {
    "current": 93,
    "isKeyWord": false,
    "tex": "n",
    "type": "StringLiteral",
    "value": "n",
  },
  {
    "current": 94,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
  {
    "current": 95,
    "isKeyWord": true,
    "tex": "]",
    "type": "RParen",
    "value": "]",
  },
]
`;

exports[`matrix with empty el > should generate ast 2`] = `
{
  "body": [
    {
      "alignment": "c",
      "dividerIndices": [],
      "lparen": "\\\\left[",
      "params": [
        [
          {
            "body": [
              {
                "tex": "1",
                "type": "Const",
                "value": "1",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "tex": "0",
                "type": "Const",
                "value": "0",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "tex": "\\\\cdots",
                "type": "Const",
                "value": "cdots",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "tex": "0",
                "type": "Const",
                "value": "0",
              },
            ],
            "type": "Flat",
          },
        ],
        [
          {
            "body": [
              {
                "body": [
                  {
                    "tex": "s",
                    "type": "Const",
                    "value": "s",
                  },
                  {
                    "params": {
                      "tex": "1",
                      "type": "Const",
                      "value": "1",
                    },
                    "tex": "_{ $1 }",
                    "type": "ParamOne",
                  },
                ],
                "type": "Flat",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "body": [
                  {
                    "tex": "b",
                    "type": "Const",
                    "value": "b",
                  },
                  {
                    "params": {
                      "tex": "11",
                      "type": "Const",
                      "value": "11",
                    },
                    "tex": "_{ $1 }",
                    "type": "ParamOne",
                  },
                ],
                "type": "Flat",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "tex": "\\\\cdots",
                "type": "Const",
                "value": "cdots",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "body": [
                  {
                    "tex": "b",
                    "type": "Const",
                    "value": "b",
                  },
                  {
                    "params": {
                      "body": [
                        {
                          "tex": "1",
                          "type": "Const",
                          "value": "1",
                        },
                        {
                          "tex": "n",
                          "type": "Const",
                          "value": "n",
                        },
                      ],
                      "type": "Flat",
                    },
                    "tex": "_{ $1 }",
                    "type": "ParamOne",
                  },
                ],
                "type": "Flat",
              },
            ],
            "type": "Flat",
          },
        ],
        [
          {
            "body": [
              {
                "tex": "\\\\vdots",
                "type": "Const",
                "value": "vdots",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "tex": "\\\\vdots",
                "type": "Const",
                "value": "vdots",
              },
            ],
            "type": "Flat",
          },
          {
            "tex": "",
            "type": "Const",
            "value": "",
          },
          {
            "body": [
              {
                "tex": "\\\\vdots",
                "type": "Const",
                "value": "vdots",
              },
            ],
            "type": "Flat",
          },
        ],
        [
          {
            "body": [
              {
                "body": [
                  {
                    "tex": "s",
                    "type": "Const",
                    "value": "s",
                  },
                  {
                    "params": {
                      "tex": "n",
                      "type": "Const",
                      "value": "n",
                    },
                    "tex": "_{ $1 }",
                    "type": "ParamOne",
                  },
                ],
                "type": "Flat",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "body": [
                  {
                    "tex": "b",
                    "type": "Const",
                    "value": "b",
                  },
                  {
                    "params": {
                      "body": [
                        {
                          "tex": "n",
                          "type": "Const",
                          "value": "n",
                        },
                        {
                          "tex": "1",
                          "type": "Const",
                          "value": "1",
                        },
                      ],
                      "type": "Flat",
                    },
                    "tex": "_{ $1 }",
                    "type": "ParamOne",
                  },
                ],
                "type": "Flat",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "tex": "\\\\cdots",
                "type": "Const",
                "value": "cdots",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "body": [
                  {
                    "tex": "b",
                    "type": "Const",
                    "value": "b",
                  },
                  {
                    "params": {
                      "body": [
                        {
                          "tex": "n",
                          "type": "Const",
                          "value": "n",
                        },
                        {
                          "tex": "n",
                          "type": "Const",
                          "value": "n",
                        },
                      ],
                      "type": "Flat",
                    },
                    "tex": "_{ $1 }",
                    "type": "ParamOne",
                  },
                ],
                "type": "Flat",
              },
            ],
            "type": "Flat",
          },
        ],
      ],
      "rparen": "\\\\right]",
      "type": "Matrix",
    },
  ],
  "type": "Root",
}
`;
